plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineWebXmlValidator.java

```diff
@@ -0,0 +1,133 @@
+import java.util.Stack;
+
+import javax.xml.parsers.ParserConfigurationException;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.wst.validation.AbstractValidator;
+import org.eclipse.wst.validation.ValidationEvent;
+import org.eclipse.wst.validation.ValidatorMessage;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXParseException;
+import com.google.cloud.tools.eclipse.appengine.facets.WebProjectUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.io.ByteStreams;
+
+import org.eclipse.wst.validation.ValidationResult;
+import org.eclipse.wst.validation.ValidationState;
+
+/**
+ * Validator for appengine-web.xml
+ */
+public class AppEngineWebXmlValidator extends AbstractValidator {
+  
+  private static final Logger logger = Logger.getLogger(
+      AppEngineWebXmlValidator.class.getName());
+  private IProject project;
+  
+  @Override
+  public void validationStarting(IProject project,
+      ValidationState state, IProgressMonitor monitor) {        
+    this.project = project;
+  }
+  
+  /**
+   * Locates appengine-web.xml in project.
+   */
+  @Override
+  public ValidationResult validate(ValidationEvent event, ValidationState state,
+      IProgressMonitor argMonitor) {
+    IResource resource = event.getResource();
+    IFile file = WebProjectUtil.findInWebInf(project, new Path("appengine-web.xml"));
+    try (InputStream in = file.getContents()) {
+      byte[] bytes = ByteStreams.toByteArray(in);
+      return validate(resource, bytes);
+    } catch (IOException | CoreException | ParserConfigurationException ex) {
+      logger.log(Level.WARNING, ex.getMessage());
+      return new ValidationResult();
+    }
+  }
+    
+  /**
+   * Validates given XML file. Assigns markers to all elements found in blacklist.
+   */
+  @VisibleForTesting
+  static ValidationResult validate(IResource resource, byte[] bytes) 
+      throws CoreException, IOException, ParserConfigurationException {
+    try {
+      Stack<BannedElement> blacklist = BlacklistSaxParser.readXml(bytes);
+      return addMessages(resource, bytes, blacklist);
+    } catch (SAXException ex) {
+      return createSaxErrorMessage(resource, ex);
+    }
+  }
+  
+  /**
+   * Adds message to project's IResource for every BannedElement in the blacklist.
+   */
+  @VisibleForTesting
+  static ValidationResult addMessages(IResource resource, byte[] bytes,
+      Stack<BannedElement> blacklist) throws IOException {
+    ValidationResult result = new ValidationResult();
+    Map<BannedElement, Integer> bannedElementOffsetMap = ValidationUtils.getOffsetMap(bytes, blacklist);
+    for (BannedElement element : bannedElementOffsetMap.keySet()) {
```

Comment:
You can also iterate over `map.entries()` instead of `keySet()` and use `entry.getKey()` and `entry.getValue()` instead of another lookup from the map to get the value.